Frameworks/UIKit/UILabel.mm

```diff
@@ -44,61 +44,70 @@ @implementation UILabel {
         return;
     }
 
-    float curFontSize = _originalFontSize;
-    _font = [_font fontWithSize:curFontSize];
-
-    CGRect rect;
-    rect = [self bounds];
-    if (rect.size.width == 0.0f || rect.size.height == 0.0f) {
+    CGRect rect = [self bounds];
+    if (rect.size.width == 0.0f || rect.size.height == 0.0f || _text == nil) {
         return;
     }
 
-    while (curFontSize > _minimumFontSize && curFontSize > 0.0f) {
-        CGSize size = CGSizeZero;
+    // MaxFitFontSize records currently found maximum font size that fits the bounds,
+    // It starts with _minimumFontSize because it is smallest font size allowed by the UILabel
+    float maxFitFontSize = _minimumFontSize;
+
+    float curFontSize = [_font pointSize];
+    // Start our trial with fonts bigger than _minimumFontSize
+    if (curFontSize < maxFitFontSize) {
+        curFontSize = 2.0 * maxFitFontSize;
+        _font = [_font fontWithSize:curFontSize];
+    }
 
-        //  A single line of text should be clipped
+    // Do binary search on maximum font size that fits the bound
+    do {
+        // Caculating rect size using font and line break settings
+        CGSize size = CGSizeZero;
         if (_numberOfLines == 1) {
-            size = [_text sizeWithFont:_font constrainedToSize:CGSizeMake(rect.size.width, 0.0f) lineBreakMode:UILineBreakModeClip];
+            size = [_text sizeWithFont:_font constrainedToSize:CGSizeMake(0.0f, 0.0f) lineBreakMode:UILineBreakModeClip];
         } else {
-            //  Multiple lines of text should be wrapped
             size = [_text sizeWithFont:_font constrainedToSize:CGSizeMake(rect.size.width, 0.0f) lineBreakMode:UILineBreakModeWordWrap];
         }
 
         if (size.width < rect.size.width && size.height <= rect.size.height) {
-            break;
+            // Update maxFitFontSize, and double current font size and retry
+            maxFitFontSize = curFontSize;
+            curFontSize *= 2.0f;
+        } else {
+            // Current font size does not fit
+            if (curFontSize - maxFitFontSize <= 1.0f) {
+                // Distance between current font size wwhich does nit fit
+                // and maximum font size which fits is less than 1.0, break out
+                break;
+            } else {
+                // Re-try with middle value between maxFitFontSize and current Font Size
+                curFontSize = (curFontSize + maxFitFontSize) / 2.0f;
+            }
         }
 
-        curFontSize -= 1.0f;
         _font = [_font fontWithSize:curFontSize];
-    }
-
-    if (curFontSize < _minimumFontSize) {
-        curFontSize = _minimumFontSize;
-    }
-    if (curFontSize < 0.0f) {
-        curFontSize = 1.0f;
-    }
+    } while (true);
 
-    _font = [_font fontWithSize:curFontSize];
+    _font = [_font fontWithSize:maxFitFontSize];
 
-    [self adjustTextLayerSize];
+    [self _adjustTextLayerSize];
 }
 
-- (void)adjustTextLayerSize {
+- (void)_adjustTextLayerSize {
```

Comment:
>_adjustTextLayerSize [](start = 8, length = 20)

nit: This name is misleading.   How about _applyXamlProperties or _updateXamlElement?